File: rs/anda_cognitive_nexus/src/kml/select.rs
//! # Choosing what a mutation acts on
//!
//! A mutation either names its target — `ARCHIVE :old` — or binds it with a
//! selection block: `ARCHIVE ?c WHERE { ?c CONCEPT {type: "Note"} } LIMIT 100`.
//! This module is the second case, and it is where KML borrows the KQL solver.
//!
//! ## The snapshot the block reads
//!
//! A selection block reads the state the **transaction started from**, not the
//! rows this transaction has staged. Two reasons, and the second is the load
//! bearing one:
//!
//! - `anda_db` has no transactional read path, so the solver reads committed
//!   rows by construction;
//! - clause order carries no mutation semantics (§24). A block that could see
//!   its own transaction's writes would make `ARCHIVE ?c WHERE {...}` mean
//!   something different depending on where it sat in the `MUTATE` — which is
//!   exactly the dependency §24 exists to deny.
//!
//! Elements this transaction minted are `pending` and no pattern matches them,
//! so the rule holds by construction rather than by a filter someone must
//! remember to write.
//!
//! ## The order a `LIMIT` cuts
//!
//! §52.7 says a bounded sweep must not be assumed deterministic *unless the
//! runtime documents an order*. This one does: ascending element id, which is
//! creation order within a kind. A sweep that runs twice therefore takes the
//! same elements both times instead of an arbitrary slice that changes with an
//! index's internal ordering.

use anda_kip::{ElementRef, KipError, Scalar, WhereClause};

use super::value::Bindings;
use crate::governance::Permission;
use crate::id::ElementId;
use crate::kql;
use crate::store::Store;
use crate::tx::Transaction;

/// The elements one clause will act on, ascending by id.
///
/// The ids are private and reachable only through [`Targets::authorized`],
/// which is the whole point of the type: a clause cannot mutate an element it
/// forgot to authorize, because it cannot get at the id without asking. The
/// two-step shape exists because resolving borrows the transaction and
/// authorizing mutates it, and collapsing them would only be possible by
/// letting the ids out first.
#[must_use = "targets have to be authorized before a clause may act on them"]
pub struct Targets {
    ids: Vec<ElementId>,
    permission: Permission,
}

impl Targets {
    /// The elements this clause may act on, ascending by id.
    ///
    /// Every target — whether the caller named it or a sweep found it — is
    /// authorized individually. A sweep does not silently skip what it may not
    /// touch: a mutation reporting success for work it did not do is the
    /// "accepted, then quietly dropped" failure this engine keeps finding.
    ///
    /// Empty is an ordinary outcome: a sweep whose block matched nothing
    /// changes nothing.
    pub async fn authorized(self, tx: &mut Transaction) -> Result<Vec<ElementId>, KipError> {
        for id in &self.ids {
            tx.authorize_element(*id, self.permission).await?;
        }
        Ok(self.ids)
    }

    /// How many elements the block bound, before authorization.
    ///
    /// Only for the statements that must bind exactly one operand and want to
    /// say so precisely; it discloses a count of elements the caller could
    /// already read, because the solver filtered them.
    ///
    /// No `is_empty` companion on purpose: "bound nothing" is not a question
    /// any clause asks, and adding the method to satisfy the shape would put
    /// back a disclosure surface with no caller.
    #[allow(clippy::len_without_is_empty)]
    pub fn len(&self) -> usize {
        self.ids.len()
    }
}

/// Resolves the targets of a mutation clause.
///
/// `WHERE` present:
/// - a `?var` target is bound by the block, and every element it takes is a
///   target (bounded by `LIMIT`);
/// - a directly named target (`:id`, `"id"`) keeps its identity and the block
///   is a **guard**: it must have at least one solution, or the clause does
///   nothing. A guard that fails is not an error — that is what makes
///   `ARCHIVE :x WHERE {...}` a conditional sweep rather than a precondition.
///
/// `WHERE` absent: the statement names its target, and `LIMIT` has nothing to
/// bound.
#[allow(clippy::too_many_arguments)]
pub async fn targets(
    store: &Store,
    tx: &Transaction,
    what: &str,
    permission: Permission,
    target: &ElementRef,
    where_clauses: Option<&Vec<WhereClause>>,
    limit: Option<&Scalar>,
    b: &Bindings<'_>,
) -> Result<Targets, KipError> {
    let Some(clauses) = where_clauses else {
        return Ok(Targets {
            ids: vec![b.element_ref(target)?],
            permission,
        });
    };

    let mut cx = kql::Context::open(
        store,
        &tx.cx.space,
        b.request,
        b.operation,
        &tx.authority,
        &tx.auth,
    )
    .await?;
    let solutions = cx.solve(clauses).await?;
    let limit = limit
        .map(|scalar| b.scalar_u64(scalar, "LIMIT"))
        .transpose()?
        .map(|limit| limit as usize);

    let variable = match target {
        // A handle this transaction already bound is a different element from
        // whatever the block selects, and silently preferring one of them would
        // make the statement mean something the author cannot see. Two names,
        // two meanings, one spelling: refuse.
        ElementRef::Handle(name) if tx.handles().contains_key(name.as_str()) => {
            return Err(KipError::reference_error(format!(
                "?{name} is bound both by a clause in this transaction and by the {what} \
                 selection block; give the selection variable a different name"
            )));
        }
        ElementRef::Handle(name) => Some(name.as_str()),
        _ => None,
    };

    let ids = match variable {
        Some(name) => {
            if !solutions.binds(name) {
                return Err(KipError::reference_error(format!(
                    "the {what} selection block does not bind ?{name}"
                )));
            }
            let mut ids = solutions.elements_of(name);
            // §52.7: a documented order, so a bounded sweep is repeatable.
            ids.sort_unstable();
            ids.dedup();
            if let Some(limit) = limit {
                ids.truncate(limit);
            }
            ids
        }
        None => {
            // A guard: the statement already names its target.
            if solutions.is_empty() {
                Vec::new()
            } else {
                vec![b.element_ref(target)?]
            }
        }
    };

    Ok(Targets { ids, permission })
}

File: ts/kip-do/src/kml/select.ts
/**
 * Selection blocks: the `WHERE` a mutation may carry.
 *
 * Three properties make this a contract rather than a convenience, and all
 * three are pinned by `fixtures/kip-conformance-2.0/mutation-selection.json`:
 *
 * **A selection block reads the state the transaction started from** (§24).
 * Clause order carries no mutation semantics, so a sweep must not see what an
 * earlier clause of the same `MUTATE` created — if it could, order would carry
 * semantics after all. This holds by construction here: an element this
 * transaction minted is `pending`, and no ordinary candidate scan matches it.
 * Explicit output handles read their final creation views, frozen before the
 * mutation pass and subject to the same read authority as stored rows.
 *
 * **`LIMIT` cuts in ascending element id.** §52.7 permits a runtime to document
 * an order, and documenting one is what makes a bounded sweep repeatable: run
 * the same capped sweep twice and it takes the same elements, rather than
 * whichever ones the storage layer felt like returning.
 *
 * **Matching nothing changes nothing.** An `UPDATE` never creates, so a block
 * that selects no elements is a `no_effect` receipt and not an error — the
 * caller asked about a set that turned out to be empty, which is an answer.
 */

import { errors } from '../errors.js'
import {
  compareElementId,
  formatElementId,
  parseElementId,
  type ElementId,
} from '../id.js'
import type { JsonMap } from '../json.js'
import type { Permission } from '../governance/index.js'
import type { ElementRef, Scalar, WhereClause } from '../kip/ast.js'
import { Context } from '../kql/context.js'
import { solveAll } from '../kql/matching.js'
import { validateWhere } from '../kql/validation.js'
import { elementBinding, type MutableSolution } from '../kql/solution.js'
import { baseline } from '../projection/policy.js'
import type { Transaction } from '../tx.js'
import { handleId, parameter, scalar, type Bindings } from './value.js'

/**
 * What a clause's target names, once resolved.
 *
 * The ids are private and reachable only through {@link Targets.authorized}, so
 * a clause cannot act on a selected element without having authorized it. A new
 * clause that forgets is a compile error rather than an ungoverned sweep.
 *
 * **A sweep that reaches something it may not touch fails.** It does not quietly
 * do less: an operation that reports success having skipped half its targets is
 * the defect shape this project keeps finding, and here it would also be a
 * disclosure — the caller could learn which elements exist outside its Grant by
 * counting what a sweep changed.
 */
export class Targets {
  /** Whether the target was named directly rather than selected. */
  readonly direct: boolean
  readonly #ids: readonly ElementId[]
  readonly #permission: Permission

  constructor(ids: readonly ElementId[], direct: boolean, permission: Permission) {
    this.#ids = ids
    this.direct = direct
    this.#permission = permission
  }

  /** How many elements were selected, without handing any of them over. */
  get size(): number {
    return this.#ids.length
  }

  /** The elements, once every one of them has been authorized. */
  authorized(tx: Transaction): ElementId[] {
    for (const id of this.#ids) tx.authorizeElement(id, this.#permission)
    return [...this.#ids]
  }
}

/**
 * Resolves a mutation's target, whether it names one element or selects a set.
 *
 * A `?variable` target is bound by the `WHERE`; an `:id` or `"id"` target names
 * the element already and needs no block (§58).
 */
export function resolveTargets(
  tx: Transaction,
  b: Bindings,
  target: ElementRef,
  where: readonly WhereClause[] | null,
  limit: Scalar | null,
  request: JsonMap | undefined,
  operation: JsonMap | undefined,
  what: string,
  permission: Permission,
): Targets {
  const direct = (id: ElementId) => new Targets([id], true, permission)
  if (where === null) {
    if ('Handle' in target) return direct(parseElementId(handleId(b, target.Handle)))
    return direct(parseElementId(namedId(b, target, what)))
  }

  const cx = new Context(tx.store, tx.env, tx.cx.space, tx.authority, tx.auth)
  const initial: MutableSolution = new Map()
  const queryBindings = {
    request: request ?? {}, operation: operation ?? {}, policy: baseline(), ambient: initial,
  }
  for (const [name, wire] of Object.entries(tx.handles())) {
    const id = parseElementId(wire)
    const view = tx.handleView(id)
    if (view !== null) cx.seedElement(id, view)
    initial.set(name, elementBinding(id))
  }
  const scope = validateWhere(cx, where, queryBindings, new Set(initial.keys()))
  if ('Handle' in target && !scope.has(target.Handle)) {
    throw errors.referenceError(`the ${what} selection block does not bind ?${target.Handle}`)
  }
  const solutions = solveAll(cx, where, [initial], queryBindings)
  const cap = limit === null ? null : count(b, limit, `${what} LIMIT`)

  if (!('Handle' in target) || tx.handle(target.Handle) !== null) {
    // A named target keeps its identity and the block is a **guard**, the way
    // the KML grammar's note 6b and `MERGE CONCEPT` already read it: no
    // solution means the clause does nothing, which is what makes
    // `TRANSITION :x TO "archived" WHERE {...}` a conditional move rather than
    // a precondition that fails. Refusing the shape instead — as this engine
    // used to — made a statement the reference engine executes a syntax error
    // here.
    return solutions.length === 0 || cap === 0
      ? new Targets([], true, permission)
      : direct(parseElementId('Handle' in target ? handleId(b, target.Handle) : namedId(b, target, what)))
  }

  const seen = new Map<string, ElementId>()
  for (const solution of solutions) {
    const bound = solution.get(target.Handle)
    if (bound?.kind !== 'element') continue
    seen.set(formatElementId(bound.id), bound.id)
  }

  // The documented order, applied before the cap: `LIMIT` says how many are
  // affected, not which, unless the runtime says which — and this one does.
  const ids = [...seen.values()].sort(compareElementId)
  return new Targets(cap === null ? ids : ids.slice(0, cap), false, permission)
}

/** The element id a `"id"` or `:parameter` target names. */
function namedId(
  b: Bindings,
  target: Exclude<ElementRef, { Handle: string }>,
  what: string,
): string {
  if ('Id' in target) return target.Id
  const value = parameter(b, target.Param)
  if (typeof value !== 'string') {
    throw errors.typeMismatch(
      `${what} needs an element id, got ${JSON.stringify(value)}`,
    )
  }
  return value
}

function count(b: Bindings, value: Scalar, what: string): number {
  const resolved = scalar(b, value)
  if (typeof resolved !== 'number' || !Number.isSafeInteger(resolved) || resolved < 0) {
    throw errors.typeMismatch(
      `${what} must be a non-negative integer, got ${JSON.stringify(resolved)}`,
    )
  }
  return resolved
}
